Emacs: use --replace-warn instead of deprecated --replace

AndersonTorres opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented:
@ofborg build emacs30.pkgs.magit emacs29.pkgs.magit emacs28.pkgs.magit

AndersonTorres commented:
@ofborg build emacs30.pkgs.magit

jian-lin commented on pkgs/applications/editors/emacs/default.nix:
Please adding the formatting commit hash to the `.git-blame-ignore-revs` file to reduce 'git blame' noise.

jian-lin commented on pkgs/applications/editors/emacs/native-comp-driver-options-30.patch:
```diff
@@ -0,0 +1,16 @@
++  (append
++   (cond ((eq system-type 'darwin) '("-Wl,-w"))
++         ((eq system-type 'cygwin) '("-Wl,-dynamicbase")))
++   '(@backendPath@)) ;;;; Added by NixOS
```
```suggestion
+   '(@backendPath@)) ;; Added by Nixpkgs
```

jian-lin reviewed (commented):
Hmm, why does this cause massive rebuilds?

jian-lin commented:
There are errors in the CI:
```
Error: error ("Can't read whole string")
  error("Can't read whole string")
  package-read-from-string("((emacs \"26.1\") (compat \"30.0.0.0\"))")
```
Similar errors occur in emacs-overlay's hydra: https://hydra.nix-community.org/eval/64935?filter=aarch64-linux.unstable.packages-unstable.&compare=64902&full=

However, I cannot reproduce that.  `(package-read-from-string "((emacs \"26.1\") (compat \"30.0.0.0\"))")` evals without an issue in my locally built Emacs 30.

jian-lin commented on pkgs/applications/editors/emacs/default.nix:
There are errors in the CI:
```
Error: error ("Can't read whole string")
  error("Can't read whole string")
  package-read-from-string("((emacs \"26.1\") (compat \"30.0.0.0\"))")
```
Similar errors occur in emacs-overlay's hydra: https://hydra.nix-community.org/eval/64935?filter=aarch64-linux.unstable.packages-unstable.&compare=64902&full=

However, I cannot reproduce that.  `(package-read-from-string "((emacs \"26.1\") (compat \"30.0.0.0\"))")` evals without an issue in my locally built Emacs 30.

jian-lin commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -136,288 +144,379 @@ let
+      gettext
+      gnutls
+      (lib.getDev harfbuzz)
+      jansson
```
`jansson` is not needed in Emacs 30.

AndersonTorres replied:
Done. In a sense.

AndersonTorres commented:
@ofborg build emacs29.pkgs.magit

AndersonTorres commented:
> Hmm, why does this cause massive rebuilds?

Some changes touch the `make-emacs` code.

jian-lin commented:
> Some changes touch the `make-emacs` code.

I skimmed through the third commit "[wip] emacs30: init" and it seems factoring out changes about `--replace-fail` into another PR targeting staging is enough to avoid the massive rebuilds of this PR.

AndersonTorres commented:
> > Some changes touch the `make-emacs` code.
> 
> I skimmed through the third commit "[wip] emacs30: init" and it seems factoring out changes about `--replace-fail` into another PR targeting staging is enough to avoid the massive rebuilds of this PR.

As I have  reading about, some of these replacements are useless, since `/bin/pwd` never appears.

I will factor it later, I will sleep now.

AndersonTorres commented:
@ofborg build emacs30.pkgs.magit

AndersonTorres commented:
@ofborg build emacs29.pkgs.magit

AndersonTorres commented:
Well, there is some change breaking things between releases here.
Or is there something happening at ofBorg?

AndersonTorres commented:
I am "duplicating" it at https://github.com/NixOS/nixpkgs/pull/342408 in order to split the probable mass-rebuild thing.

AndersonTorres commented:
Is this still mass-rebuilding?

AndersonTorres commented:
nixpkgs-review says `6257 packages updated`.
This is just a file being renamed, what is happening?

Please do not tell me this is a reproducible-source-paths issue...
https://nix.dev/guides/best-practices.html#reproducible-source-paths

jian-lin commented:
TIL: `builtins.path`

With this patch applied, the output hash of `emacs` does not change,

```diff
diff --git a/pkgs/applications/editors/emacs/make-emacs.nix b/pkgs/applications/editors/emacs/make-emacs.nix
index f995b073d377..01b5b6fd2c2f 100644
--- a/pkgs/applications/editors/emacs/make-emacs.nix
+++ b/pkgs/applications/editors/emacs/make-emacs.nix
@@ -158,9 +158,12 @@ mkDerivation (finalAttrs: {
 
   patches = patches fetchpatch ++ lib.optionals withNativeCompilation [
     (substituteAll {
-      src = if lib.versionOlder finalAttrs.version "29"
-            then ./native-comp-driver-options-28.patch
-            else ./native-comp-driver-options.patch;
+      src = builtins.path {
+        path = if lib.versionOlder finalAttrs.version "29"
+               then ./native-comp-driver-options-28.patch
+               else ./native-comp-driver-options-29.patch;
+        name = "native-comp-driver-options.patch";
+      };
       backendPath = (lib.concatStringsSep " "
         (builtins.map (x: ''"-B${x}"'') ([
           # Paths necessary so the JIT compiler finds its libraries:

```

AndersonTorres commented:
Now, I will use this branch to refactor Emacs expression with mass-rebuilds.

AndersonTorres commented:
@ofborg build emacs30.pkgs.compat

AndersonTorres commented:
@ofborg build emacs30-pgtk.pkgs.magit

AndersonTorres commented:
I will postpone the hammering to Master branch.

AndersonTorres commented:
Let's use homeopatic pull requests.
This is just to change replace to replace-fail.

jian-lin commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -185,20 +185,20 @@ mkDerivation (finalAttrs: {
+    ''
       for makefile_in in $(find . -name Makefile.in -print); do
-        substituteInPlace $makefile_in --replace /bin/pwd pwd
+        substituteInPlace $makefile_in --replace-fail /bin/pwd pwd
```
This makes Emacs 28 (and maybe more) fail to build because not all founded makefiles have /bin/pwd.  You can run `find . -name Makefile.in -print0 | xargs --verbose -0 -I '{}' -n 1 rg /bin/pwd '{}'` in the Emacs 28.2 source code repo to verify this.

AndersonTorres replied:
Solved.

jian-lin commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -185,20 +185,20 @@ mkDerivation (finalAttrs: {
       done
     ''
-
-    ""
```
nit(non-blocking): removing this empty string is not part of `use --replace-fail`.

jian-lin commented on pkgs/applications/editors/emacs/make-emacs.nix:
```diff
@@ -185,20 +185,20 @@ mkDerivation (finalAttrs: {
-
+        --replace-fail /usr/share/locale ${gettext}/share/locale
+    ''
+  ] ++ lib.optionals (lib.versionOlder version "29") [
```
nit(non-blocking): instead of adding an optional list, the local pattern is to use `optionalString` in the list.

AndersonTorres committed:
emacs: use `--replace-warn`

`--replace-fail` can fail sometimes, given that not all files contain
the string being replaced.
